src/include/processor/operator/partitioner.h

```diff
@@ -69,28 +67,40 @@ struct PartitionerLocalState {
 };
 
 struct PartitioningInfo {
-    DataPos keyDataPos;
-    std::vector<DataPos> columnDataPositions;
-    std::vector<common::LogicalType> columnTypes;
+    common::idx_t keyIdx;
     partitioner_func_t partitionerFunc;
 
-    PartitioningInfo(DataPos keyDataPos, std::vector<DataPos> columnDataPositions,
-        std::vector<common::LogicalType> columnTypes, partitioner_func_t partitionerFunc)
-        : keyDataPos{keyDataPos}, columnDataPositions{std::move(columnDataPositions)},
-          columnTypes{std::move(columnTypes)}, partitionerFunc{std::move(partitionerFunc)} {}
+    PartitioningInfo(common::idx_t keyIdx, partitioner_func_t partitionerFunc)
+        : keyIdx{keyIdx}, partitionerFunc{std::move(partitionerFunc)} {}
     inline std::unique_ptr<PartitioningInfo> copy() {
-        return std::make_unique<PartitioningInfo>(keyDataPos, columnDataPositions,
-            common::LogicalType::copy(columnTypes), partitionerFunc);
+        return std::make_unique<PartitioningInfo>(keyIdx, partitionerFunc);
     }
 
     static std::vector<std::unique_ptr<PartitioningInfo>> copy(
         const std::vector<std::unique_ptr<PartitioningInfo>>& other);
 };
 
+struct PartitionerDataInfo {
+    std::vector<common::LogicalType> columnTypes;
+    std::vector<std::unique_ptr<evaluator::ExpressionEvaluator>> columnEvaluators;
+    std::vector<bool> defaultColumns;
+
+    PartitionerDataInfo(std::vector<common::LogicalType> columnTypes,
+        std::vector<std::unique_ptr<evaluator::ExpressionEvaluator>> columnEvaluators,
+        std::vector<bool> defaultColumns)
+        : columnTypes{std::move(columnTypes)}, columnEvaluators{std::move(columnEvaluators)},
+          defaultColumns{std::move(defaultColumns)} {}
+    inline std::unique_ptr<PartitionerDataInfo> copy() {
```

Comment:
I think you can use object rather than unique_ptr. If so use `EXPLICIT_COPY_DEFAULT_MOVE` macro.